src/util/utility-types.ts

```diff
@@ -4,3 +4,23 @@
  * Required since React.ComponentProps doesn't think it's a JSXElementConstructor/ReactElement. https://github.com/FB-PLP/traject/pull/11929/files#r948531375
  */
 export type ExtractProps<T> = T extends React.ComponentType<infer P> ? P : T;
+
+/**
+ * Given two types, only allow one of them to be valid
+ */
+export type Only<T, U> = {
+  [P in keyof T]: T[P];
+} & {
+  [P in keyof U]?: never;
+};
+
+/**
+ * Given two types, specify that only one of them can be in
+ * the structure at a given time
+ */
+export type Either<T, U> = Only<T, U> | Only<U, T>;
+
+/**
+ * Given two types, do the same as Either type, OR possibly both
+ */
+export type EitherBoth<T, U> = Either<T, U> | (T & U);
```

Comment:
Naming is hard. This conceptually seems similar to inclusive-or, and the others are exclusive-or.

Reply:
`Either` is currently XOR, and `EitherBoth` is OR. Pretty CS-y names. I think those above could work, more consistent with each other for sure

Reply:
Right, exactly. EitherBoth is inclusive, and Either is exclusive.

I like the idea of using more CS-y names. Makes it more clear from the name what to expect.